paramiko/pipe.py

```diff
@@ -46,25 +46,50 @@ def __init__(self):
         self._closed = False
 
     def close(self):
-        os.close(self._rfd)
-        os.close(self._wfd)
-        # used for unit tests:
         self._closed = True
 
+        # We cannot do anything about closing errors. It is only a
+        # "best effort" approach
+        try:
+            os.close(self._rfd)
+        except:
+            pass
+        try:
+            os.close(self._wfd)
+        except:
+            pass
+
     def fileno(self):
         return self._rfd
 
     def clear(self):
         if not self._set or self._forever:
             return
+
+        # .read() does not need to handle a race condition with .close()
+        # because the pipe is created, cleared and closed from the same
+        # "client thread". The "server thread" sets the pipe and only .set()
+        # suffers from a race condition with .close()
         os.read(self._rfd, 1)
         self._set = False
 
     def set(self):
         if self._set or self._closed:
             return
         self._set = True
-        os.write(self._wfd, b"*")
+
+        # This try fixes a race condition with .close()
+        # 1. The write/server thread sees ._closed == False and continues
+        # 2. The close/client thread closes the descriptors before the
+        #    write/server thread writes.
+        # 3. The write/server fails to write() because the FD has been closed
+        try:
+            os.write(self._wfd, b"*")
+        except OSError as e:
+            if e.errno == 9 and self._closed:
```

Comment:
I'd suggest using `errno.EBADF` with `errno` module in `e.errno == 9`.

Reply:
Done